### PR TITLE
#70: Create a Pre Commit Git Hook Preventing Commits to master

### PR DESCRIPTION
**Pull request by amrwc**
#70

Screenshot 2020-01-26 at 7 13 13 pm

### CODE REVIEW

File: bin/setup.ps1 [hunk 1]
```diff
@@ -0,0 +1 @@
+# TODO:
```
Comment (faibz) [inline bin/setup.ps1]: nice

**Comment by faibz** [comment]
What advantage does this give over simply making master a protected branch in GitHub?

**Comment by amrwc** [comment]
> What advantage does this give over simply making master a protected branch in GitHub? I understand that this seems to be pre-commit so could point something out before attempting a push, but seems less clear than the functionality GitHub provides (visually that is).
> 
> This idea seems like it'd be beneficial for bigger projects that more people working on this, but I'm struggling to see the benefit for us right now.
> 
> Edit: I've made an example branch "test" with branch protections.

The advantage is that it prevents committing to a wrong branch locally. Guarding such branch on GitHub is also recommended. It's just for convenience and it's the right tool for the job.

I'm thinking about rewriting both the `setup` script and `pre-commit` hook in Python to make one script for multiple platforms and thought that maybe in the future it will be worth having a single setup script that may or may not do some more complicated tasks, such as setting up the local environment (say, dockerised SQL Server).

But again, it's nothing critical, I just thought it would be nice to have.

**Comment by faibz** [comment]
Okay, if you think it provides some value and is optional, then I'm happy with it.

I definitely support a script or at least a README addition for setting up dependencies. Would make it way easier to get started. Worth noting that Docker Desktop needs Hyper-V on Windows which isn't present in Windows 10 Home, but can always get Docker Toolbox, a VM, or WSL.

**Pull request closed without merging**